### ISSUE TITLE
kids-teacher: barge-in over-cancels real responses (mid-thought pauses + tool-confirm echo)

### ISSUE THREAD

**Issue by theroadrunnershow** [theroadrunnershow/myra-language-teacher#56]
## Summary

Live Reachy sessions show the assistant getting cut off mid-response in two distinct patterns. Pulled and analysed `~/myra.log` from two sessions (~02:20–02:23 and ~02:34–02:44) — diagnosis below.

The two failure modes are **independent root causes** and need separate fixes:

- **Problem 1 — mid-thought pause truncation** (this issue, fix landing now)
- **Problem 2 — echo/noise barge on short tool-confirmations** (kept open in this issue)

---

## Problem 1 — `silence_duration_ms=500` truncates mid-thought pauses

**Symptom**: kid speaks → pauses ~0.5–1s mid-sentence → Gemini's VAD ends the user turn → robot starts replying → kid resumes their original thought → resumption trips `interrupted=True` and cancels the (premature) reply. User experiences a fragmented exchange where one input + one response should have happened.

**Evidence — session 2, turn 7→8** (`/tmp/myra-session/myra-2.log`):
```
02:39:23  child partial: 'I want you to remember that when you talk to Myra,
                          make sure that you sneakily a'                 ← truncated
02:39:25  assistant:     'I can definitely remember some things,'        ← premature reply
02:39:25  server reported interrupted=True                                ← kid resumed
02:39:25  cancel_response invoked                                         ← reply killed
02:39:27  child partial: 'lessons from my life'                          ← rest of sentence
```
The full utterance was almost certainly *\"…sneakily ask **<pause>** lessons from my life\"*. Gemini's `silence_duration_ms=500` ended the turn during the pause.

**Fix (this PR/commit)**: bump `_VAD_SILENCE_DURATION_MS` from 500 → 1000 in `src/kids_teacher_gemini_backend.py`. Single-line config change. Trade: ~500ms extra latency at end of *real* turns, acceptable for a 4yo speech cadence.

Tests updated: `tests/test_kids_teacher_gemini_backend.py:244`.

---

## Problem 2 — short tool-confirmations get barged within 1–2s (still open)

**Symptom**: after a tool call (e.g. `remember_face`), the model emits a short confirmation (\"Got it, I'll remember Mike next time!\") and is interrupted within ~1.5s — *with no following child transcript* for ≥2s. Pattern repeats across multiple back-to-back `remember_face` calls.

**Evidence — session 1, turns 6 & 7** (`/tmp/myra-session/myra.log`):
| Turn | When | Last partial heard | Cause |
|---|---|---|---|
| 6 | 02:22:33 → 02:22:34 | `\"Got it - I'll remember Abi next time! Do you want to tell me one special story\"` | `interrupted=True` → cancel; no child transcript follows |
| 7 | 02:22:40 → 02:22:41 | `\"Got it! I'll remember Mike next time! Are there any\"` | same |

Likely cause: speaker→mic echo bleed (no hardware AEC on the Reachy mic) tripping Gemini's server VAD on the robot's own confirmation audio. Possibly compounded by ambient noise being misclassified (cf. garbled transcript `\"die neue City U11\"` earlier in the same session).

### Reference: how pollen-robotics handles barge-in

Reviewed [`pollen-robotics/reachy_mini_conversation_app`](https://github.com/pollen-robotics/reachy_mini_conversation_app). They:
- Rely entirely on server-side VAD (OpenAI: `ServerVad(interrupt_response=True)`; Gemini: built-in)
- Flush player queue + reset wobbler on any interrupt
- **No AEC, no min-utterance gate, no echo protection, no tool-confirmation special case**

Their reference robot must have enough mic/speaker isolation that this doesn't trigger. Borrowing from their code won't fix it for us — we need gates they don't have.

### Options explored (no LLM, no network — local-only)

Decided against an ollama-based barge classifier: latency too high (`memory_reconciler` is already timing out at default 30s in our logs — bad signal for hot-path use) and LLMs only see text, not the audio signals that matter for echo detection.

Better stack, layered:

1. **L0 grace window (instant)** — ignore barge in first ~600ms of any assistant turn. Track first-audio-chunk timestamp in `kids_teacher_robot_bridge.py`.
2. **L1 acoustic echo correlator (~5–10ms)** — numpy cross-correlation between speaker outgoing buffer and mic incoming buffer over a 0–200ms delay search. High correlation → echo, ignore the barge. Both buffers already pass through the bridge process. numpy already in deps via faster-whisper.
3. **L2 speculative pause (300ms wall)** — on `speech_started`, *pause* the player (don't drain). If no transcript delta arrives within 300ms, *resume*. Pause is reversible; cancel is not.
4. **L3 rule-based score (<1ms)** — features: `string_similarity_to_assistant_recent` (rapidfuzz, already a dep), `transcript_word_count`, `mic_rms / speaker_rms`, `contains_interrupt_word` (\"stop\", \"wait\", \"no\", child name), `time_since_speak_started_ms`, `was_tool_confirmation_turn`. Hand-tuned weights → cancel/resume.
5. **L4 (later, after data)** — replace L3 weights with a tiny scikit-learn `LogisticRegression` trained on logged events. Auto-label using \"did the kid re-ask within 10s of cancel\" as ground truth. <1ms inference.

Also drop the redundant transcript-delta cancel path at `kids_teacher_realtime.py:274–275` — it's more aggressive than pollen and contributes to over-cancelling.

### Instrumentation that pays for itself

Whatever we ship, log per decision:
```
event=barge_decision verdict=resume|cancel reason=L1_echo|L2_no_transcript|L3_score
features={time_ms: 540, sim: 0.71, mic_rms: 0.012, speaker_rms: 0.18, words: 0, ...}
correction_observed_within_10s=True|False
```
This turns every session into training data for free.

### Suggested sequence for Problem 2

- [ ] Verify speaker + mic buffers are co-located in the bridge process (precondition for L1)
- [ ] Drop the redundant transcript-delta barge path (`kids_teacher_realtime.py:274–275`)
- [ ] Ship L0 + L1 + L2 + L3 as one commit
- [ ] Add per-decision logging
- [ ] Re-validate against next live session; check whether tool-confirmation cancels disappear
- [ ] L4 (learned classifier) only if rule-based plateau is hit

---

## Out of scope (separate concerns surfaced during analysis)

- **Process supervision**: kids-teacher is launched foreground via `python … 2>&1 | tee ~/myra.log` with no `nohup` / systemd. Process died silently at 02:44:29 in session 2 — bash on `pts/0` survived but child Python is gone, no traceback in log, no OOM. Anything spoken after that point got nothing. Track separately.
- **`memory_reconciler` ollama timeouts** at 02:21:46, 02:23:02, 02:39:58 — non-fatal but the local LLM is unreliable.
- **Motion stack `apply_skip ≈ 60/heartbeat` while `gain_state=robot_speaking`** — robot likely looks frozen during long utterances. Separate UX bug.
